[FEATURE] WIP: Make rigid body simulation fully differentiable

SonSang opened the pull request:
<!--  Thanks for sending a pull request!  Please:

1. Follow our contributor guidelines: 
   https://github.com/Genesis-Embodied-AI/Genesis/blob/main/.github/CONTRIBUTING.md
2. Prepare your PR according to the "Submitting Code Changes" 
   section of https://github.com/Genesis-Embodied-AI/Genesis/blob/main/.github/CONTRIBUTING.md
3. Provide a concise summary of your changes in the Title above
4. Prefix the title according to the type of issue you are addressing. Use:
    - [BUG FIX] for non-breaking changes which fix an issue
    - [FEATURE] for non-breaking changes which add functionality
    - [MISC] for minor changes such as improved inline documentation or fixing typos
    - [CHANGING] for changes that will change simulation's behaviour
    - [BREAKING] **in addition to the above** for breaking changes, i.e., a fix or feature that would cause existing APIs or functionality to change
-->

## Description
<!--- Describe your changes in detail -->

This PR aims at stitching differentiable components in the rigid body simulation to make it fully differentiable end-to-end. Also, it will provide a minimal demo to demonstrate the differentiability of the rigid body simulation.

Specifically, this PR would contain:
* Fully differentiable rigid body simulation
* Minimal demo for differentiable rigid body simulation (e.g. solving optimization problem)
* More informative error handlings for differentiable simulation for users

## Related Issue
<!--- This project only accepts pull requests related to open issues.
If suggesting a new feature or change, please discuss it in an issue first.
If fixing a bug, there should be an issue describing it with steps to reproduce.

Please link to the relevant issue here, e.g. via `Resolves <issue-url>`.
Cf. https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue -->

Resolves Genesis-Embodied-AI/Genesis#2069

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been / Can This Be Tested?
<!--- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

TBD

## Screenshots (if appropriate):

TBD

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [ ] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [ ] I tested my changes and added instructions on how to test it for reviewers.

<!--- Optionally -->
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

SonSang committed:
migrated add inequality functions in constraint solver

SonSang committed:
stash

SonSang committed:
Merge branch 'main' into diffrigiddemo

SonSang committed:
make grad zero in reset_grad

duburcqa commented on genesis/engine/solvers/rigid/rigid_solver_decomp.py:
```diff
@@ -299,6 +302,9 @@ def build(self):
 
+            if getattr(self._options, "sparse_solve", False):
+                gs.raise_exception("Sparse solve is not supported yet when requires_grad is True.")
+
```
It would be nice to have unit tests to catch all this forbidden configurations.

SonSang committed:
1. added one another rigid adjoint cache to check integrity of the backward pass
2. fixed bug in save_ckpt, needed to set copy=True in converting taichi field to numpy array

SonSang committed:
1. use dofs force as differentiable input
2. use backward mode for rigid body sim only when computing gradient

SonSang committed:
1. added grad window
2. stabilized differentiable contact, now can solve simple optimization problem

SonSang committed:
stash

SonSang committed:
1. accelerated backward pass for some functions by pulling out static inner loops outside
2. fixed some geometric functions in utils.geom.py to avoid nan grads
3. added rigid entity state queries (e.g. qpos, dofs_vel)

SonSang committed:
stash

duburcqa commented:
Superseded by https://github.com/Genesis-Embodied-AI/genesis-world/pull/2842
